source/common/config/delta_subscription_state.cc

```diff
@@ -14,52 +14,130 @@ namespace Config {
+    }
+    ASSERT(!requested_resource_state_.contains(r));
     // Ideally, when interest in a resource is added-then-removed in between requests,
     // we would avoid putting a superfluous "unsubscribe [resource that was never subscribed]"
     // in the request. However, the removed-then-added case *does* need to go in the request,
     // and due to how we accomplish that, it's difficult to distinguish remove-add-remove from
     // add-remove (because "remove-add" has to be treated as equivalent to just "add").
     names_added_.erase(r);
-    names_removed_.insert(r);
+    if (actually_erased) {
+      names_removed_.insert(r);
+      in_initial_legacy_wildcard_ = false;
+    }
+  }
+  // If we unsubscribe from wildcard resource, drop all the resources that came from wildcard from
+  // cache. Also drop the ambiguous resources - we aren't interested in those, but we didn't know if
+  // those came from wildcard subscription or not, but now it's not important any more.
+  if (cur_removed.contains(Wildcard)) {
+    wildcard_resource_state_.clear();
+    ambiguous_resource_state_.clear();
   }
 }
 
 // Not having sent any requests yet counts as an "update pending" since you're supposed to resend
 // the entirety of your interest at the start of a stream, even if nothing has changed.
 bool DeltaSubscriptionState::subscriptionUpdatePending() const {
-  return !names_added_.empty() || !names_removed_.empty() ||
-         !any_request_sent_yet_in_current_stream_ || must_send_discovery_request_;
+  if (!names_added_.empty() || !names_removed_.empty()) {
+    return true;
+  }
+  // At this point, we have no new resources to subscribe to or any
+  // resources to unsubscribe from.
+  if (!any_request_sent_yet_in_current_stream_) {
+    // If we haven't sent anything on the current stream, but we are actually interested in some
+    // resource then we obviously need to let the server know about those.
+    if (!requested_resource_state_.empty()) {
+      return true;
+    }
+    // So there are no new names and we are interested in nothing. This may either mean that we want
+    // the legacy wildcard subscription to kick in or we actually unsubscribed from everything. If
+    // the latter is true, then we should not be sending any requests. In such case the initial
+    // wildcard mode will be false. Otherwise it means that the legacy wildcard request should be
+    // sent.
+    return in_initial_legacy_wildcard_;
+  }
+
+  // At this point, we have no changes in subscription resources and this isn't a first request in
+  // the stream, so even if there are no resources we are interested in, we can send the request,
+  // because even if it's empty, it won't be interpreted as legacy wildcard subscription, which can
+  // only for the first request in the stream. So sending an empty request at this point should be
+  // harmless.
+  //
+  // If sending empty requests at this point is actually harmful, we would need to add "&&
```

Comment:
Should this be deleted given that we establish this is safe above?

Reply:
Oh, here I wanted to get a second opinion here, but I forgot to ask about it explicitly. Sorry about that. But yeah, in my opinion sending empty non-initial requests is harmless, even if rather wasteful, so I'll drop this paragraph.